### PR TITLE
Fix: Wrong list of users on search page

### LINKED ISSUE #7670: [BUG] Wrong list of users on search page when query is involved
### Description

Upon loading the page with the search query "?keyword=typescript," the default users list is presented before implementing the query as a filtering mechanism.

### Screenshots

https://github.com/EddieHubCommunity/LinkFree/assets/70102539/c0e8ccd5-ad68-4c18-9fe6-d63132aef8f2

### Additional information

I think the correct thing is to display the filtered user directly.

### PR DESCRIPTION
**Pull request by llii0046**
## Fixes Issue
Closes #7670

## Changes proposed

The issue originally stemmed from the implementation of the `getStaticProps` function, which populated the user list with an array of randomly selected users. Upon detecting a search keyword within the router query, an asynchronous fetchUsers(keyword) request would be initiated on the client-side to update the user list.

To rectify this situation, I've adopted a **server-side solution** and implemented the following changes:

1. I've replaced the `getStaticProps` with `getServerSideProps`. The former lacks the capability to access the query object which is necessary in this case.

2. I've encapsulated some of the logic into function, thereby enhancing the readability and maintainability of the code.

## Check List (Check all the applicable boxes) 

- [x] My code follows the code style of this project.
- [ ] My change requires changes to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] This PR does not contain plagiarized content.
- [x] The title of my pull request is a short description of the requested changes.

## Screenshots

https://github.com/EddieHubCommunity/LinkFree/assets/100179608/5ac7a057-267a-4b14-83f0-f51fcf72f6f9

## Note to reviewers

<a href="https://gitpod.io/#https://github.com/EddieHubCommunity/LinkFree/pull/8514"></a>

### CODE REVIEW

File: pages/search.js [hunk 1]
```diff
@@ -11,44 +10,73 @@ import logger from "@config/logger";
...
+      serverProps.filteredUsers = await fetchUsersByKeyword(keyword);
+    }
+    else {
+      serverProps.randUsers = await fetchRandomUsers();
+    }
+  } catch (e) {
+    logger.error(e, "ERROR fetching users");
   }
 
+  serverProps.tags = await fetchTags();
```
Comment (azhechkova) [inline pages/search.js:65]: I think it may be better to move `fetchTags` to the try block too

Comment (xkrishguptaa) [reply pages/search.js:65]: It is already under a try/catch block in the above code

Comment (azhechkova) [reply pages/search.js:65]: sorry, haven't noticed :)

File: pages/search.js [hunk 2]
```diff
@@ -11,44 +10,73 @@ import logger from "@config/logger";
...
-  const pageConfig = config.isr.searchPage; // Fetch the specific configuration for this page
-  let data = {
-    tags: [],
-  };
-  let users = [];
-  try {
-    users = await getProfiles();
-  } catch (e) {
-    logger.error(e, "ERROR search users");
+async function fetchUsersByKeyword(keyword) {
```
[the remark this answers is not in the document]
Comment (llii0046) [reply pages/search.js:15]: I agree

[the remark this answers is not in the document]
Comment (llii0046) [reply pages/search.js:15]: Hi, Do you think it's a good idea to create a folder called `profiles` under services then move these functions into there? It not only maintains the original structure, but also separates the service logic of specific functions.

[the remark this answers is not in the document]
Comment (azhechkova) [reply pages/search.js:15]: hi, I think that's a good idea, as it separates the logic and structures it well

[the remark this answers is not in the document]
Comment (eddiejaoude) [reply pages/search.js:15]: Good suggestion but for now as it is only used in this file, maybe keep it in a separate function but the same file

**Review by eddiejaoude — APPROVED** [review]
Looks good, thank you 👍

**Pull request merged**